### ISSUE TITLE
Submit form normally with preprocessing steps

### ISSUE THREAD

**Issue by zach-waggoner** [davidkpiano/react-redux-form#886]
### The Problem

I need my form to be submitted in the normal way, but I also need to add some asynchronous preprocessing first (I need to upload a file to S3 and submit the url of the uploaded file with the form). To accomplish this, I have tried to utilize the `pending` state of the form. When the user submits the form, `pending` is false, so I call `event.preventDefault()`, upload the file to S3, set `pending` to `true`, and then dispatch the `submit` action. This calls the submit handler a second time with `pending` set to true, so the handler doesn't call `event.preventDefault()`. The problem is that even though my handler isn't calling `event.preventDefault()` the second time, and even though the `action` is specified on the `Form`, the form isn't actually submitting. This is also the case if the second submission happens by calling `document.getElementById('form-id').submit()` instead of `dispatch(actions.submit('form-model'))`.

### Reproducible Code Example

[https://codepen.io/anon/pen/VWojNP](https://codepen.io/anon/pen/VWojNP)

**Comment by davidkpiano** [comment]
So this seems to be a separate issue outside of RRF - in short, you're using `submit` in a way that wasn't intended.

The real solution for this sounds like you just need to have a promise in `handleSubmit` that continues (`.then`) with the URL, append that URL to your form data, and submit it (again, in `.then(...)`.

**Comment by zach-waggoner** [comment]
I don't think I understand. I _am_ appending the URL to the form data and resubmitting it. The problem is that the resubmission doesn't work. I have verified that it does work in a normal HTML form onSubmit handler — see https://codepen.io/anon/pen/JJgaNb. If you click submit, you will see `preprocessing...` logged in the console, followed by the form disappearing and a couple of console errors, indicating the form was submitted to the phony action I added to it. In the code example in my original comment (which I have updated to use a promise for the sake of clarity), clicking on submit causes `preprocessing...` to be logged in the console, but the form is never actually submitted, even though the submit handler is called again.
